### ISSUE TITLE
React router unmounts, set state and change language not playing well together

### ISSUE THREAD

**Issue by johnraz** [i18next/react-i18next#796]
**Describe the bug**
When changing language (with XHR backend setup), changing the route in react-router and setting a state value I get the well-known warning below:

> Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.

I'm guessing it's due to the async XHR call to the language file because it only happens on the first load.

I tried to go around the issue but couldn't nail it so far.

I set up a code sandbox example here: https://codesandbox.io/s/6vm8o90jpk

Note: I'm not sure this belongs to this repo or not, if you want me to move it somewhere else please let me know.

Thanks for your help!

**Occurs in react-i18next version**
10.5.1

**To Reproduce**

1. Go to 'https://6vm8o90jpk.codesandbox.io/en/hello'
2. Open the console
3. Click on 'Change lang'
4. See  the error in console

**Expected behaviour**

**Screenshots**
http://recordit.co/tKnLmlyKMg

![](http://g.recordit.co/tKnLmlyKMg.gif)

**OS :**

- Device: MBP 2017 15"
- Browser: Chrome 72.0.3626.121

**Comment by jamuhl** [comment]
not sure what we should do with this? I mean the issue is in your code as you set state on the changeLanguage callback...

guess this

```js
  function changeLanguage() {
    i18n
      .changeLanguage(i18n.language === "en" ? "fr" : "en", () => {
        history.push(`/${i18n.language}${t("/hello")}`);
      })
      .then(() => {
        setData("State");
      });
  }
```

does not belong into rendering logic of a component...you change the route and expect this component still to exist to set a state....

**Comment by johnraz** [comment]
Hi @jamuhl,

Thanks so much for the feedback, I think you finally made it click for me!

I'm still learning how to handle react state and react-router properly and I took a "static / backend route" approach when thinking about this which lead me to put my routes way to high in the stack and to not split component between routes properly. 

So my language switcher ended up within a route with the actual "page" component while I should instead keep it outside of that route on a separate "always matching" route. 

This is a rookie react mistake...
Thanks for petting my ignorance and taking the time to answer this one seriously.

Closing this.

**Comment by johnraz** [comment]
So I made some more testing and the issue is not fixed by moving the language switcher outside of the route...

Another more advanced code sandbox is available here: 
https://codesandbox.io/s/ov583v6099

A recording of the behavior:
http://recordit.co/QuGzFkQ59D

I'm not going to re-open, @jamuhl just let me know if you consider this as non-related to react-i18next, I'll post this on the react spectrum anyway.

Cheers.

**Comment by jamuhl** [comment]
I only see one issue your code is:

```js
  const changeLanguage = nextLang => {
    i18n.changeLanguage(nextLang).then(() => {
      console.log(i18n.language, t("/hello"));
      history.push(`/${i18n.language}${t("/hello")}`);
    });
  };
```

but history.push(`/${i18n.language}${t("/hello")}`); will use the default namespace translation to lookup the key -> changing it to history.push(`/${i18n.language}${t("common:/hello")}`); specifing you have the translation on "common" namespace is all to do

**Comment by jamuhl** [comment]
https://www.i18next.com/translation-function/essentials#accessing-keys-in-different-namespaces

**Comment by johnraz** [comment]
I tried loading the default common namespace in a new code sandbox but the problem persists. I’ll see if I get some idea out of spectrum.

**Comment by jamuhl** [comment]
made a fork: https://codesandbox.io/s/6yzqv37v8z

works with the change i said... do i miss sometning not working?

**Comment by johnraz** [comment]
For me, loading the page and using the drop down menu  to change language triggers the error in your fork. Make sure to not use the buttons on top but the dropdown menu. I made it that way to compare a working simple button vs a broken advanced drop down.

**Comment by jamuhl** [comment]
Ah now i see...haha material UI -> not even saw that there is a dropdown...lol

Well not sure...seems more an problem that comes from combination of history.push and material UI...not much we can do here.

**Comment by durasj** [comment]
I am currently having a similar problem with the hooks and it looks like the `useTranslation` hook will set the internal state after the component is no longer mounted. Looking at the code, this looks very suspicious: https://github.com/i18next/react-i18next/blob/61c4645a900ffa018deb9789b5ec390e12181980/src/useTranslation.js#L71

In case the load will finish after the component is no longer mounted, this will lead to error. This loading is not canceled in the cleanup of useEffect.

**Comment by jamuhl** [comment]
@durasj yes...please create a new issue for this...not yet sure how to solve this...the hook is not aware of the component mounted or not...guess setState hook should be safe by it's design...will see what we can do

**Comment by durasj** [comment]
@jamuhl I will try to come up with a good reproducible example over the weekend and create the issue. Well, the setState hook won't technically fail, but it can't prevent you from holding a reference to it through your resetT() method. The naive way to solve this would be to do something like:
```js
const mounted = true;

function resetT() {
    mounted && setT({ t: i18n.getFixedT(null, namespaces[0]) });
}

useEffect(() => {
    ...

    // unbinding
    return () => {
        mounted = false;
        ...
    };
});
```
But it would be much better if there would be a way to cancel the namespace loading. So the `loadNamespaces` would return cancel method that could be called on unmount.

**Comment by jamuhl** [comment]
@durasj just googled a little...guess the code you appended is the best solution. loadNamespaces is deep inside i18next core - so chances are small we can add a cancel to that.

**Comment by johnraz** [comment]
@jamuhl:
This is not related to an issue between react-router's history.push and material-ui.

I edited my code sandbox example to add an option in the dropdown to redirect to a page that doesn't trigger a language change but only a `history.push` and everything works as expected.

The code sandbox: https://codesandbox.io/s/yj5k978wq9

Video of the reproduced bug: http://g.recordit.co/HujCNQH8Z6.gif

I'm about to drop the idea of having a language switcher that doesn't do a hard reload in the browser which is kinda sad but I really can't find a proper fix.

One way I find that made it work is declaring my route's path as:
`/:lang(en|fr)/:page(english-name|french-name|other-language-name)/` and then it works because it doesn't seem to unmount the route in react-router.

As a side note, regarding the cancel, this blog post explains how to cancel http requests... I guess this only applies to the XHR backend but well I figured it could help https://medium.freecodecamp.org/how-to-work-with-react-the-right-way-to-avoid-some-common-pitfalls-fc9eb5e34d9e#d5d7

**Comment by jamuhl** [comment]
Cancelling would be not easy to get right...as we do not only have xhr backends...so a solution checking if the component using the hook will be the way to go...still i guess not so easy to get that inside the hook.

**Comment by johnraz** [comment]
I guess a generic cancel method on the backend could be the way to go. I’m probably overlooking a lot of details though but this is what first comes to my mind.

**Comment by jamuhl** [comment]
loadNamespace calls get deduplicated asserting a namespace load request gets triggered only once while asserting callback is called for all calls on the right time...not so simple as it seems on the first sight.

so while you might cancel a loadNamespace - there is a chance there is a pending request for some other component...so open for a PR...but my guess it's more work than needed.

**Comment by jamuhl** [comment]
setting

```js
,
    react: {
      bindI18n: false
    }
```

might also be an options -> as rerender based on language change is already triggered by router

**Comment by rares-lupascu** [comment]
maybe u already solved it but
https://codesandbox.io/s/7oxwnn74q

does this solve ur issue?

**Comment by johnraz** [comment]
@jamuhl: using `bindI18n` to `false` is not working because then the route definition is not updated when the language change, i18next's rerender will trigger the root route's rerender which in turns will re-mount the component.

@rares-lupascu setting `useSuspense` to `false` did indeed make things better. I'm not 100% sure what other implication that might have. @jamuhl, can you detail what that setting changes that could make the situation better?

**Comment by jamuhl** [comment]
@rares-lupascu fork seems to work at least when changing route map to full locales en-US: https://codesandbox.io/s/kx5z2wz005

**Comment by jamuhl** [comment]
useSuspense false -> you do not use suspense -> so while loading translations there is no fallbackUI but an additional flag ready to render or not.

Might be you do not use suspense or try moving suspense on a page level component inside the router...idk

**Comment by johnraz** [comment]
Hi @jamuhl,

I'm going to re-open this after further investigations this morning.
I finally could create a very minimal use-case scenario here: https://codesandbox.io/s/z2znn0qxy4

To trigger the warning, just hit the `fr` button after loading the page and you should see it in the console.
As before, the warning is only showed the first time the `fr` button is clicked, any subsequent clicks are fine.

I'm not setting any state manually to trigger the issue anymore.
Material UI is not involved anymore.

Here is another code sandbox where the "Hello" component doesn't use react-i18n and doesn't trigger the warning: https://codesandbox.io/s/p5vw2636x0

I believe this shows that something is happening in react-i18n with state management that triggers the warning as when no translation is used, there is no warning.

Really sorry to come back with this without an actual solution but I'm really stuck at what could be causing it 😞

**Comment by jamuhl** [comment]
My guess is the Problem is on react-router side or it is just not meant to be used in that way.

The problem is caused by using Suspense outside of the `BrowserRouter` -> disabling suspense on the `AppComp` more or less solves it:

```js
const rootElement = document.getElementById("root");
ReactDOM.render(
  <Suspense fallback="loading...">
    <App useSuspense={false} />
  </Suspense>,
  rootElement
);
```

if not doing so when chaning to `fr` i18next will load translations for french and trigger the Suspense.

So my suggestion for getting this to work:
- Set useSuspense false on the AppComponent
- Move the Suspense Component to a page level component or inside the BrowserRouter

eg.: https://codesandbox.io/s/98v150owp

**Comment by johnraz** [comment]
Setting `useSuspense` to false + using ready or moving the Suspense component within the router is not solving it.

The example you linked above triggers the warning twice.

I will try to post this on react-router's issues tracker as well.

Thanks!

**Comment by jamuhl** [comment]
Ok...eventual this time ;)

changeLanguage change:

```js
function LanguageSwitcherComp(props) {
  const { children, i18n, t, history, match } = props;
  const changeLanguage = nextLang => {
    // const newUrl = localizeLink(routeName, match.params);
    const newUrl = `/${nextLang}/${t(match.params.page)}`;
    // debugger;
    console.log(newUrl);
    history.push(newUrl);
    i18n.changeLanguage(nextLang);
  };
  const handleClickSelectLanguage = (event, index) => {
    changeLanguage(supportedLanguages[index]);
  };
  return (
   //....
```

=> not trigger history.push in the callback

https://codesandbox.io/s/qz1o2z7804

**Comment by johnraz** [comment]
Ah! It seems to be fixed indeed!
Thinking about it again now --> calling `history.push` is enough, we don't need to call `i18n.changeLanguage` as the language will be taken from the URL anyway!

So in your last example, `i18n.changeLanguage` is probably never actually called, right?
I guess the problem would still be there if we did translate the path alone, without prepending with `/fr` or `/en` and then having to rely on `i18n.changeLanguage`.

This is not what I intend to do right now but maybe it would be good to investigate? What do you think?

**Comment by johnraz** [comment]
Also in your last example, removing the redirect (which I initially put there for conveniency) shows that the newUrl and history.push should happen after the changeLanguage not before.

See: https://codesandbox.io/s/92mp3w1z54

**Comment by jamuhl** [comment]
// const newUrl = localizeLink(routeName, match.params);
    const newUrl = `/${nextLang}${t(match.params.page || "/hello", {
      lng: nextLang
    })}`;

**Comment by johnraz** [comment]
Sadly this won't work as the `"/hello"` has to be translated.

The `changeLanguage` has to be completed before building the new path and doing the `history.push`. I fear there is no way around it...

Isn't there a way to "wait" so that i18n.changelanguage is completely done before triggering the history.push so that no more set state activity is ongoing?

**Comment by jamuhl** [comment]
changeLanguage callback is correctly called when translations are loaded

point is HOC and hook work independent of that doing there own stuff -> triggering Suspense on the currently rendered component.

My guess is https://github.com/i18next/react-i18next/blob/master/src/useTranslation.js#L85 causing the problem -> not sure if we could safely remove that as the tree gets newly rendered from the Suspense component downwards...

**Comment by jamuhl** [comment]
At this point i would say this is a chicken - egg problem...would have to debug proper through it...and see which line of code triggers the issue - but currently not enough time

**Comment by ghardin137** [comment]
I went through the useTranslation hook code and found these two event handlers being created but never removed during cleanup

https://github.com/i18next/react-i18next/blob/master/src/useTranslation.js#L60

**Comment by johnraz** [comment]
@ghardin137 isn't it done here?: https://github.com/i18next/react-i18next/blob/219230c92eafa9ee3cdbc7761a1f9041089bb799/src/useTranslation.js#L60-L68

**Comment by johnraz** [comment]
We discussed this with @ghardin137 on discord yesterday and he suggested to use `useLayoutEffect` for the cleanup of the event handlers in hope it would trigger at the "right time".

I'll try that ASAP but I need some time to find my way through the library development process etc, as already said I'm not super at ease with JS libs dev and need to get up to speed with that first.

**Comment by jamuhl** [comment]
Ok debugged through the code. Not sure why but somehow there is a component getting unmounted before the unbind of subscriptions. Made a little change in react-i18next@10.6.1 which seems to solve this....

**Comment by johnraz** [comment]
@jamuhl so far my tests shows that it's indeed fixed, thanks a ton for your time and availability 👍

**Comment by jamuhl** [comment]
If you like this module don’t forget to star this repo. Make a tweet, share the word or have a look at our https://locize.com to support the devs of this project -> there are many ways to help this project :pray:

**Comment by johnraz** [comment]
I just wanted to add for further people visiting this that using `i18n.changeLanguage` outside of effect hook is probably not a good idea as now that this issue is fixed, I'm facing other issues by not doing so.

If you have to change language before navigating, as I wanted to do here, it's better to have a local state in your language switcher, set that local state to the "next language" value when the UI is used and then when the component is re-rendered, in a `useEffect` or `componentDidMount/componentDidUpdate` compare that state value with the actual value of `i18n.language` and if it differs then you can use `history.push` with the translated path.
